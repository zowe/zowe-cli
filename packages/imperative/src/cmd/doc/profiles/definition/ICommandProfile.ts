/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

/**
 * Profile load specifications - causes profiles to auto-load for this command and will cause options to be
 * exposed on the command for the user to specify the desired profile by name.
 * Profiles specified by the user on the command line always take precedence over defaults.
 * @export
 * @interface ICommandProfile
 */
export interface ICommandProfile {
    /**
     * An array of profile "types" that are required for the execution of this command.
     * Imperative will automatically attempt to load (the default OR the profile specified by name on the CLI).
     * If a profile of a required type cannot be found, the command will fail.
     * @type {string[]}
     * @memberof ICommandProfile
     */
    required?: string[];
    /**
     * An array of profile "types" that are optional for the execution of this command.
     * Imperative will automatically attempt to load (the default OR the profile specified by name on the CLI).
     * If a profile of an optional type cannot be found, the command can still continue. It is the responsiblity of the
     * handler (for which this profile specification is defined) to properly handle an optional profile not be present
     * when the handler is invoked.
     * @type {string[]}
     * @memberof ICommandProfile
     */
    optional?: string[];
    /**
     * An array of profile "types" to suppress automatic generation of options on the command. By default, Imperative
     * will generate an CLI option in the form "--<type>-profile" for each required & optional profile specified on this
     * object.
     * @type {string[]}
     * @memberof ICommandProfile
     */
    suppressOptions?: string[];
}
