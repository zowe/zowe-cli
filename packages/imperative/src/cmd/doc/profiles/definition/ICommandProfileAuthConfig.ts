/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IPartialCommandDefinition } from "../../IPartialCommandDefinition";

export interface ICommandProfileAuthConfig {
    /**
     * Name of the authentication service
     */
    serviceName: string;

    /**
     * Path to the handler for the authentication service.
     * The handler should inherit from Imperative BaseAuthHandler.
     */
    handler: string;

    /**
     * Command properties for `auth login <serviceName>`
     */
    login?: IPartialCommandDefinition;

    /**
     * Command properties for `auth logout <serviceName>`
     */
    logout?: IPartialCommandDefinition;
}
