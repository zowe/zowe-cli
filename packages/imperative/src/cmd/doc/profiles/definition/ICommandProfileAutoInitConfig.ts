/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IPartialCommandDefinition } from "../../IPartialCommandDefinition";

export interface ICommandProfileAutoInitConfig {
    /**
     * Path to the handler for the authentication service.
     * The handler should inherit from Imperative BaseAuthHandler.
     */
    handler: string;

    /**
     * The provider giving connection information and details for the automatically generated config
     */
    provider: string;

    /**
     * Command properties for `config auto-init`
     */
    autoInit?: IPartialCommandDefinition;

    /**
     * The type of profile associated with the provider, if any
     */
    profileType?: string;
}
