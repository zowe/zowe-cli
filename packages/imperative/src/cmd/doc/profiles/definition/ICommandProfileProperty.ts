/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IProfileProperty } from "../../../../profiles";
import { ICommandOptionDefinition } from "../../option/ICommandOptionDefinition";
/**
 * Extended version of a team profile schema property that can include option definitions
 * for auto-generated commands.
 * @export
 * @interface ICommandProfileProperty
 * @extends {IProfileProperty}
 */
export interface ICommandProfileProperty extends IProfileProperty {
    /**
     * This option definition will be used to auto-generate profile commands.
     * This is the same type used by normal Imperative command definitions.
     */
    optionDefinition?: ICommandOptionDefinition;
    optionDefinitions?: ICommandOptionDefinition[];
}
