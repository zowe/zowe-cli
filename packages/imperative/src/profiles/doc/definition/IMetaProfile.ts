/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IProfileTypeConfiguration } from "../config/IProfileTypeConfiguration";

/**
 * The meta profile is saved in each of the profile "type" directories under the profile manager root directory.
 * The meta profile is always named "<type>_meta" and is always stored in YAML format. The contents contain the
 * default profile for the type AND the configuration document for the type. The configuration document is persisted
 * to allow external callers of the Profile Manager APIs (outside of an Imperative init'd environment) to load/save
 * profiles (useful for programmatic invocation from editor plugins, such as VS Code).
 * @export
 * @interface IMetaProfile
 */
export interface IMetaProfile<T extends IProfileTypeConfiguration> {
    /**
     * The default profile for the "type" - this profile will be loaded by the manager if the default is requested.
     * This is a convience for users of the CLI (and programmatic invocation) to avoid having to supply a profile
     * name for the type. You can also set the default using the Profile Manager APIs.
     *
     * @memberof IMetaProfile
     */
    defaultProfile: string;
    /**
     * The Configuration Document supplied for this type.
     * @type {T}
     * @memberof IMetaProfile
     */
    configuration: T;
}
