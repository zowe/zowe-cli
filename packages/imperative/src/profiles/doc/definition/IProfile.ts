/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

/**
 * Profiles are a user configuration mechanism built into the Imperative framework and profiles are intended to be
 * exploited on commands that require large sets of configuration items (more than would be feasible on the command
 * line). In addition, profiles allow users to dynamically and easily switch configuration for the invocation of a
 * particular command. One common use case for profiles, is storing "connection details" (host, user, API version,
 * port, credentials, etc.) required by a command to communicate with a remote instance/service.
 *
 * Another feature of profiles are "dependencies". A profile can "point to" any number of additional profiles, which
 * will be loaded when the original is loaded. A common use case for this feature is if you need "connection
 * details" (to communicate with a remote instance), but the unique serivce you are invoking also requires additional
 * input/configuration. You can "point to", via the dependencies, the "connection details" profile.
 *
 * Profiles also help you segregate/isolate portions of your CLI (if it is complex enough to warrent). For example,
 * a common pattern is to organize your CLI via "groups", where each group is isolated and requires seperate
 * configuration.
 *
 * Profiles are loaded (sometimes automatically, by definition on your commands) for the invocation of that command.
 * Meaning, they are only loaded when needed/required.
 *
 * See the "ProfileManager" for profile management details.
 *
 * @export
 * @interface IProfile
 */
export interface IProfile {
    /**
     * The list of profile that this profile is dependent on. Can be explicitly named (for tight  coupling of
     * interdependent profiles). Profile dependencies are useful to help complex CLIs segregate/isolate profile (and
     * for example, command groups) configurations, while allowing for reuse of configuration where needed. A profile
     * can have any number of dependencies of any type.
     * @type {string}
     * @memberof IProfile
     */
    dependencies?: Array<{
        /**
         * The name of the profile this profile is dependent on - profile+type uniquely qualify a profile.
         * @type {string}
         */
        name?: string;
        /**
         * The type of the profile this profile is dependent on - profile+type uniquely qualify a profile.
         * @type {string}
         */
        type: string;
    }>;
    /**
     * Indexable.
     */
    [key: string]: any;
}
