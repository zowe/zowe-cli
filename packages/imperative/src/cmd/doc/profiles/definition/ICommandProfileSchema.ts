/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandProfileProperty } from "./ICommandProfileProperty";
import { IProfileSchema } from "../../../../profiles";
/**
 * Externally exposed version of the profile schema with command options for auto-generated commands
 * @export
 * @interface ICommandProfileSchema
 * @extends {IProfileSchema}
 */
export interface ICommandProfileSchema extends IProfileSchema {
    /**
     * Command version of the properties field on the schema
     */
    properties: {
        /**
         * General mapping of property name to an ICommandProfileProperty object.
         * ICommandProfileProperty is the same as IProfileProperty except
         * that it can contain option definitions for auto-generated commands
         */
        [key: string]: ICommandProfileProperty,
    };
}
