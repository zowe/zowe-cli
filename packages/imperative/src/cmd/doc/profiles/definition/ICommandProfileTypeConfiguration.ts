/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandProfileSchema } from "./ICommandProfileSchema";
import { IProfileTypeConfiguration } from "../../../../..";
import { ICommandExampleDefinition } from "../../ICommandExampleDefinition";
import { ICommandProfileAuthConfig } from "./ICommandProfileAuthConfig";

/**
 * Cmd packages additions to the profile manager type configuration document. Used by the CliProfileManager. Allows
 * profiles to be built from command arguments. See the "CliProfileManager" for more information.
 * @export
 * @interface ICommandProfileTypeConfiguration
 * @extends {IProfileTypeConfiguration}
 */
export interface ICommandProfileTypeConfiguration extends IProfileTypeConfiguration {
    /**
     * A handler module which Imperative will require().
     * The module's default export should be a  handler that calls
     * appendResponseObject on the provided commandParameters.response
     * You do NOT have to implement writing the profile to disk -- you only have to produce
     * the final profile object that you would like to be written.
     *
     *
     * This is only required if finished profile can't be created directly from the arguments, e.g.
     * if you have --user and --password and need to always transform it into a basic auth
     *
     *
     * If omitted, Imperative will just write all fields present from the schema into the profile
     * without requiring a module
     *
     * @type {string}
     * @memberof IProfileTypeConfiguration
     */
    createProfileFromArgumentsHandler?: string;
    /**
     * The module's default export should be a handler that calls appendResponseObject on the provided
     * commandParameters.response You do NOT have to implement writing the profile to disk -- you only have to produce
     * the final profile object that you would like to be written.
     *
     * This is only required if finished updated profile can't be created directly from the arguments, e.g.
     * if certain fields that the user might specify mean that other fields should be deleted or updated.
     *
     * If omitted, Imperative will load the old profile, overwrite any fields specified by the user,
     * and write the updated profile to disk.
     *
     * @type {string}
     * @memberof IProfileTypeConfiguration
     */
    updateProfileFromArgumentsHandler?: string;
    /**
     * Examples to be displayed in the help text for the auto generated create profile command.
     *
     * @type {ICommandExampleDefinition[]}
     * @memberof IProfileTypeConfiguration
     */
    createProfileExamples?: ICommandExampleDefinition[];

    /**
     * Examples to be displayed in the help text for the auto generated update profile command.
     *
     * @type {ICommandExampleDefinition[]}
     * @memberof IProfileTypeConfiguration
     */
    updateProfileExamples?: ICommandExampleDefinition[];

    /**
     * The JSON schema document. The schema document provides a way to enforce the contents of a profile. The schema
     * conforms exactly to the JSON schema specification. You must supply all properties you would like validated
     * for correctness on the schema, except for "dependencies". Dependency schema checking is generated automatically
     * if you populate the dependencies property of this document.
     *
     * @type {ICommandProfileSchema}
     */
    schema: ICommandProfileSchema;

    /**
     * Configuration for authentication services to associate with this profile type.
     */
    authConfig?: ICommandProfileAuthConfig[];
}
