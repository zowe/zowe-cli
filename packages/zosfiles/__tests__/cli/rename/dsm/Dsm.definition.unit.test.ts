/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import i18nEnglish from "../../../../src/cli/-strings-/en";
import { DsmDefinition } from "../../../../src/cli/rename/dsm/Dsm.definition";

describe("DsDefinition", () => {
    it("should use the correct object for strings", () => {
        const strings = i18nEnglish.RENAME.ACTIONS.DATA_SET_MEMBER;

        expect(DsmDefinition.description).toBe(strings.DESCRIPTION);
        expect(DsmDefinition.children).toBeUndefined();
        expect(DsmDefinition.profile.optional).toEqual(["zosmf"]);
        expect(DsmDefinition.options).toBeUndefined();
        expect(DsmDefinition.examples).toMatchSnapshot();
        expect(DsmDefinition.description).toBe(strings.DESCRIPTION);
    });
});
