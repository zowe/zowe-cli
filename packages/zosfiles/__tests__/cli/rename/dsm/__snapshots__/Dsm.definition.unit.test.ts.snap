// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DsDefinition should use the correct object for strings 1`] = `
Array [
  Object {
    "description": "In the data set 'USER.DATA.SET', rename the member named 'MEM1' to 'MEM2'.",
    "options": "\\"USER.DATA.SET\\" \\"MEM1\\" \\"MEM2",
  },
]
`;
